### PR TITLE
Add support for custom delimiters and template contexts

### PR DESCRIPTION
**Pull request by mprahl**
The custom delimiter support will allow distinguishing between multiple kinds of templates in
the same YAML file (e.g. `{{` vs `{{hub`).

Additionally, the template context support will allow the user to use the context of the policy being resolved in their templates. For example, they may use `.ClusterName` as a variable in their templates to reference the target cluster name that this template is being generated for.

### CODE REVIEW

**Comment by mprahl** [comment]
@ckandag FYI, it's easier to review the PR commit by commit.

File: pkg/templates/templates.go
Comment (ckandag) [inline pkg/templates/templates.go:31]: ClusterID was added here.
how to set that value?  we need to determine if   we can support that ,  if processes that call this api  like policy propagator  have access to the value and  provide that info  to the processor ?

File: pkg/templates/templates.go
Comment (ckandag) [inline pkg/templates/templates.go:45]: there should be a way for  the processes to be able to set the delimiters as initial\global  configuration just like the kube-client.

So it isnt set everytime and on everyone of the calls that need it.

There needs to be two sets of methods , one takes the delims and uses that and another that doesn't take the value and uses the global if available and fails if not .

go probably has a way to specify a parameter optionally

File: pkg/templates/templates.go
Comment (ckandag) [inline pkg/templates/templates.go:113]: doesnt this change break config policy controller ?   you need to provide an option where caller is not required to set it. managedcluster dont set it

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:31]: I thought I understood that Randy required it as part of architecture call, so I haven't yet looked how to provide that, but it seems that we will need to. This is why it is here.

On the config-policy-controller side, this would just not be set.

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:113]: It wouldn't break it since `config-policy-controller` is pinned to the commit currently in the main branch. When it's ready to absorb this change, it'll have to upgrade its `go.mod` entry and change the way it calls this code.

Go doesn't have optional parameters. So perhaps the idea I had in the previous comment of making a struct containing the public API methods would work.

File: pkg/templates/templates.go
Comment (ckandag) [reply pkg/templates/templates.go:31]: While propagator  already has the ClusterName available as part of placement Decision, its doesnt already have ClusterId.  So the hesistance to just put it in already without figuring out  the backend.

It isnt actually needed for the team's  usecase either , the   request was  more like  might as well add that also along with clustername if needed later on.


I guess  you can leave it in their  for now ... depending on if we end up adding clusterId  or not I  might change this later.

**Comment by ckandag** [comment]
I am ok approving this for a first pass, but would request to hold off on updating go.mod  in config policy controller just yet.

As the propagator side of the changes are done and api here is solidified  , then cofig-policy-controller can be changed to the new api.

File: pkg/templates/templates.go
Comment (ckandag) [reply pkg/templates/templates.go:113]: Consider a user of this util pkg , who wants to resolve templates in his yaml ( not necessarily policy yaml but any yaml).
ClusterId or ClusterName may not be something that make sense in that user's case.

In such a case , it appears  user is required to pass a context with unnecessary variables set to empty. Would be nice if the user doesnt have to set it at all if its not needed for their case.


but  its alright for now I guess as its only being used by the policy components currently

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:31]: Good point. I will take it out and if we find an easy way to add this, we can add this back in later.

**Comment by mprahl** [comment]
> @mprahl
> 
> I am ok approving this for a first pass, but would request to hold off on updating go.mod in config policy controller just yet.
> 
> As the propagator side of the changes are done and api here is solidified , then cofig-policy-controller can be changed to the new api.

I will refactor this a bit using structs and I'll see if this is cleaner.

**Comment by mprahl** [comment]
@ckandag, I made some changes. Please review again.

In the first commit, I refactored the `templates` package to encapsulate the template resolution logic in a struct (`templateResolver`). This enables us to pass in a `Configuration` struct when instantiating the struct using the `New` function rather than use any global variables. In addition, the `New` function performs validation on the input configuration, where as the previous approach of global variables did not.

So usage without error handling is as follows:
```go
config := templates.Config{KubeConfig: k8Cfg}
tr, _ := templates.New(&k8Client, config)
processedMap, _ := tr.ResolveTemplate(configPolicyMap)
```

Or:

```go
config := templates.Config{KubeAPIResourceList: k8RsrcList}
tr, _ := templates.New(&k8Client, config)
processedMap, _ := tr.ResolveTemplate(configPolicyMap)
```

The second and third commit add the features needed for hub-side templates.

The usage without error handling is as follows:
```go
config := templates.Config{KubeConfig: k8Cfg}
tr, _ := templates.New(&k8Client, config)
processedMap, _ := tr.ResolveTemplate(configPolicyMap)
```

Or:

```go
config := templates.Config{
	ClusterName: "cluster0001",
	KubeConfig: k8Cfg,
	StartDelim: "{{hub",
	StopDelim:  "hub}}",
}
tr, _ := templates.New(&k8Client, config)
processedMap, _ := tr.ResolveTemplate(configPolicyMap)
```

File: pkg/templates/templates.go
Comment (JustinKuli) [inline pkg/templates/templates.go:105]: When I originally saw this line I thought "oooh, can users put other stuff into the template?". Is `.ClusterName` the only thing they might want? If not, should the config just take a whole context map, and use it here?

I understand keeping the config smaller, and not allowing arbitrary things. I'm just curious if this was brought up already.

File: pkg/templates/templates.go
Comment (JustinKuli) [inline pkg/templates/templates.go:44]: This config is not too big yet, so the approach here is fine. Just FYI, if it needs to grow more you could consider the approach outlined here: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

**Review by JustinKuli — COMMENTED** [review]
Looks good. Comments, but no actions required. I'll approve if you're done with it, just let me know :)

File: pkg/templates/templates.go
Comment (JustinKuli) [reply pkg/templates/templates.go]: > Go doesn't have optional parameters

True, although you can get something similar with variadic functions/arguments

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:70]: Good idea!

File: pkg/templates/templates_test.go
Comment (mprahl) [reply pkg/templates/templates_test.go:108]: Since each test run is parallel, we need to redeclare `test` so that each parallel test gets the correct value for `test` and not what `test` happens to be in the range loop at the time of test execution.

The go linting on this repo actually catches this which is nice.

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:105]: It's a good point. @ckandag, what do you think?

The benefit is that the caller has full control of the context that is available during template execution. This would also not require us to update `go-template-utils` every time we want to add a new field.

The downside is that `.ClusterName` is cleaner syntax and what has been approved as opposed to `.["clusterName"]`.

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:44]: That is pretty elegant in some ways but hard to understand for the API consumer. I think I prefer the Config struct approach but could be easily swayed.  :smile:

File: pkg/templates/templates.go
Comment (ckandag) [reply pkg/templates/templates.go:105]: I would think,  Ideally util package provides functionality thats common across all consumers of the pkg while any consumer specific details  like the  context or rbac etc are handled on the consumer side.   To the util pkg , the input and output  is just some valid yaml string , the contents in it - what  Kubernetes CR - shouldn't matter at all.


But we definitely need the syntax  to be  .ClusterName,    if there is a way to achieve that while allowing the consumer to pass in the context   without hardcoded it in util pkg ( via reflection or such)  that would be nice.  If not, we can live with this for now.

Also , as a consumer e.g. config-policyc-controller   the kubeclient, delimiters are one time settings that are generally common through out all  invocations of the  ResolveTemplate  call.   However, the context  is different for each invocation , just as  the  template yaml string  or the result of calling fromSecret or lookup could be different.

I really dont think the context should be part of config but should be passed optionally as param to ResolveTemplate  function.

File: pkg/templates/templates.go
Comment (ckandag) [reply pkg/templates/templates.go:44]: I referred to passing optional arguments to functions  somewhere earlier in this review , I guess the technical term I was looking for was variadic function.. phew.

But I think what we have now works just  fine. We can fine tune as we go if needed.

File: pkg/templates/templates.go
Comment (mprahl) [reply pkg/templates/templates.go:105]: Let me have a go with reflection and see if I can come up with a way to do this.

**Comment by mprahl** [comment]
@ckandag could you please take another look? Here is the diff:
https://github.com/open-cluster-management/go-template-utils/compare/1fb3464dff8a59601929437d2bbefaf1e4d0671e..66d20a3b5508ba3fd8cb879f0a9957a8483dffd4

**Comment by ckandag** [comment]
> @ckandag could you please take another look? Here is the diff:
> https://github.com/open-cluster-management/go-template-utils/compare/1fb3464dff8a59601929437d2bbefaf1e4d0671e..66d20a3b5508ba3fd8cb879f0a9957a8483dffd4

@mprahl  This looks good, exactly as hoped,  just one more thing - 

Could you consider changing the HasTemplate method also to a call on the TemplateResolver struct.  like  ResolveTemplate
If not ,  change the input params to just be the delimiters ,   the rest of the fields in the Config resource are not relevant and so  there is no-need/confusing/lacking-clarity to make users pass the config struct.

**Comment by mprahl** [comment]
> > @ckandag could you please take another look? Here is the diff:
> > https://github.com/open-cluster-management/go-template-utils/compare/1fb3464dff8a59601929437d2bbefaf1e4d0671e..66d20a3b5508ba3fd8cb879f0a9957a8483dffd4
> 
> @mprahl This looks good, exactly as hoped, just one more thing -
> 
> Could you consider changing the HasTemplate method also to a call on the TemplateResolver struct. like ResolveTemplate
> If not , change the input params to just be the delimiters , the rest of the fields in the Config resource are not relevant and so there is no-need/confusing/lacking-clarity to make users pass the config struct.

I decided to make it accept a start delimiter since I think it'd be useful to not need to instantiate the TemplateResolver if there are no templates in the string.

Could you please take another look?

**Review by ckandag — APPROVED** [review]

**Pull request merged**
